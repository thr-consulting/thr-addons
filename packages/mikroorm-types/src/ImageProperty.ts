import {Type, ValidationError} from '@mikro-orm/core';

export class ImageProperty extends Type {
	convertToDatabaseValue(value: any): any {
		try {
			if (!value) return value;
			if (!value.data || !value.timestamp || !value.type) throw new Error();
			return JSON.stringify(value);
		} catch (err) {
			throw ValidationError.invalidType(ImageProperty, value, 'JS');
		}
	}

	convertToJSValue(value: any): any {
		try {
			if (!value) return value;
			return JSON.parse(value);
		} catch (err) {
			throw ValidationError.invalidType(ImageProperty, value, 'database');
		}
	}

	getColumnType() {
		return 'json';
	}
}
