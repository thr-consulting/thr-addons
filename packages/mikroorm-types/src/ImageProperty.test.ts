import {ImageProperty} from './ImageProperty';

describe('ImageProperty', () => {
	it('should convert Image values', () => {
		const image = {
			data: 'data',
			timestamp: 'timestamp',
			type: 'type',
		};

		const prop = new ImageProperty();
		expect(prop.convertToDatabaseValue(image)).toMatchSnapshot();
		expect(prop.convertToJSValue(JSON.stringify(image))).toMatchSnapshot();
		expect(prop.getColumnType()).toMatchSnapshot();
	});

	it('should throw errors on invalid values', () => {
		const prop = new ImageProperty();

		function convertDbValue() {
			const image = 'not an image';
			prop.convertToDatabaseValue(image);
		}
		function convertJsValue() {
			const image = 'not an image';
			prop.convertToJSValue(image);
		}
		function invalidData() {
			const image = {
				data: null,
				timestamp: 'timestamp',
				type: 'type',
			};
			prop.convertToDatabaseValue(image);
		}
		function invalidTimestamp() {
			const image = {
				data: 'data',
				timestamp: null,
				type: 'type',
			};
			prop.convertToDatabaseValue(image);
		}
		function invalidType() {
			const image = {
				data: 'data',
				timestamp: 'timestamp',
				type: null,
			};
			prop.convertToDatabaseValue(image);
		}

		expect(convertDbValue).toThrowErrorMatchingSnapshot();
		expect(convertJsValue).toThrowErrorMatchingSnapshot();
		expect(invalidData).toThrowErrorMatchingSnapshot();
		expect(invalidTimestamp).toThrowErrorMatchingSnapshot();
		expect(invalidType).toThrowErrorMatchingSnapshot();
	});
});
