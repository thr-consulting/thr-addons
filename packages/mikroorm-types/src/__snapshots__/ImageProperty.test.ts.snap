// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageProperty should convert Image values 1`] = `"{\\"data\\":\\"data\\",\\"timestamp\\":\\"timestamp\\",\\"type\\":\\"type\\"}"`;

exports[`ImageProperty should convert Image values 2`] = `
Object {
  "data": "data",
  "timestamp": "timestamp",
  "type": "type",
}
`;

exports[`ImageProperty should convert Image values 3`] = `"jsonb"`;

exports[`ImageProperty should throw errors on invalid values 1`] = `"Could not convert JS value 'not an image' of type 'string' to type ImageProperty"`;

exports[`ImageProperty should throw errors on invalid values 2`] = `"Could not convert database value 'not an image' of type 'string' to type ImageProperty"`;

exports[`ImageProperty should throw errors on invalid values 3`] = `"Could not convert JS value '[object Object]' of type 'object' to type ImageProperty"`;

exports[`ImageProperty should throw errors on invalid values 4`] = `"Could not convert JS value '[object Object]' of type 'object' to type ImageProperty"`;

exports[`ImageProperty should throw errors on invalid values 5`] = `"Could not convert JS value '[object Object]' of type 'object' to type ImageProperty"`;
